### PR TITLE
fix pasting of markdown containing links

### PR DESCRIPTION
**Pull request by susanwere**
👋 This is a second attempt to fix https://github.com/github/issues/issues/5429

After merging this PR in master we realised that the fix didn't solve the issue. I did a second check and found the following issue which causes the bug.

While creating the tree walker, we specified that we only want to show nodes that are elements. This works well in the chrome browser because it generates `span` for texts that don't have a html tag as shown in the screenshots below. For Firefox, texts that don't have a html tag are skipped because they are not identified as elements thereby causing the bug.

Screenshot 2024-02-23 at 13 55 09
Screenshot 2024-02-23 at 13 55 35

The fix is to SHOW_ALL nodes for firefox and only SHOW_ELEMENTs for chrome.

Before
-----

https://github.com/github/paste-markdown/assets/25789307/2dd42c02-4118-4661-840b-491b075025ea

After
-----

https://github.com/github/paste-markdown/assets/25789307/8a1582c6-42a3-4277-99e9-e4db1b726ecc

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -383,43 +383,6 @@ describe('paste-markdown', function () {
       assert.include(textarea.value, tableMarkdown)
     })
 
-    it('pastes markdown with links correctly when identical labels are present', function () {
```
Comment (susanwere) [inline test/test.js:386]: removing these tests since we run our tests in ChromeHeadless mode and they should test firefox. The chrome context has been covered by other tests in this file.

File: src/paste-markdown-html.ts [hunk 2]
```diff
@@ -62,14 +63,11 @@ function convertToMarkdown(plaintext: string, walker: TreeWalker): string {
...
   while (currentNode && index < NODE_LIMIT) {
     index++
+    const currentNodeText =
+      navigator.userAgent.indexOf('Firefox') > -1 ? (currentNode as Text) : (currentNode.firstChild as Text)
     const text = isLink(currentNode)
       ? (currentNode.textContent || '').replace(/[\t\n\r ]+/g, ' ')
-      : (currentNode.firstChild as Text)?.wholeText || ''
-
-    // update value of markdownIgnoreBeforeIndex with current index if the current node is not a link
-    if (!isLink(currentNode)) {
```
Comment (susanwere) [inline src/paste-markdown-html.ts:70]: reverting changes from the first PR

File: src/paste-markdown-html.ts [hunk 3]
```diff
@@ -37,7 +37,8 @@ function onPaste(event: ClipboardEvent) {
   // Generate DOM tree from HTML string
   const parser = new DOMParser()
   const doc = parser.parseFromString(textHTMLClean, 'text/html')
-  const walker = doc.createTreeWalker(doc.body, NodeFilter.SHOW_ELEMENT, node =>
+  const whatToShow = navigator.userAgent.indexOf('Firefox') > -1 ? NodeFilter.SHOW_ALL : NodeFilter.SHOW_ELEMENT
```
Comment (keithamus) [inline src/paste-markdown-html.ts:40]: We should avoid user agent sniffing as it causes problems people use custom user agents. Instead we should use feature detection to determine the value of this boolean.

File: src/paste-markdown-html.ts [hunk 4]
```diff
@@ -62,14 +63,11 @@ function convertToMarkdown(plaintext: string, walker: TreeWalker): string {
   // Walk through the DOM tree
   while (currentNode && index < NODE_LIMIT) {
     index++
+    const currentNodeText =
+      navigator.userAgent.indexOf('Firefox') > -1 ? (currentNode as Text) : (currentNode.firstChild as Text)
```
Comment (keithamus) [inline src/paste-markdown-html.ts:67]: Same here. This should not use a user-agent check.

File: test/test.js (hunk 1 above)
Comment (keithamus) [reply test/test.js:386]: We can set up the tests to also run in Firefox.

**Review by keithamus — CHANGES_REQUESTED** [review]
Thanks for taking the time to make this PR @susanwere!

I've left some comments, because I think we need to go in a slightly different direction here. Let's try to ensure that:

 1. Any behaviour we add is well tested. 
 2. We don't rely on the user agent to determine which behaviour we use.

Given the above two, I would be happy to re-review this if it:

- Retains the tests that are in place (even if they're altered).
- Adds new tests (to cover the new edge cases we've discovered).
- Removes the user agent sniffing and uses feature detection instead.

If you need any assistance with this please let me know, and I'd be happy to pair.

**Review by keithamus — APPROVED** [review]
This LGTM :+1: nice work!
